apps/designer/app/designer/features/sidebar-left/panels/pages/settings.tsx

```diff
@@ -0,0 +1,136 @@
+import {
+  IconButton,
+  Button,
+  Box,
+  Label,
+  TextField,
+  useId,
+  styled,
+  Flex,
+  toast,
+  InputErrorsTooltip,
+} from "@webstudio-is/design-system";
+import { useFetcher, type FetcherWithComponents } from "@remix-run/react";
+import { ChevronDoubleLeftIcon } from "@webstudio-is/icons";
+import type { ZodError } from "zod";
+import { BaseHeader } from "../../lib/header";
+import { useEffect, useRef, useState } from "react";
+import { type Page } from "@webstudio-is/project";
+
+const Group = styled(Flex, {
+  marginBottom: "$3",
+  gap: "$2",
+  defaultVariants: { direction: "column" },
+});
+
+export type CreatePageData =
+  | { ok: true; page: Page }
+  | { errors: string | ZodError["formErrors"] };
+
+export const NewPageSettings = ({
+  onClose,
+  onSuccess,
+  onFetcherStateChange,
+  projectId,
+}: {
+  onClose?: () => void;
+  onSuccess?: (page: Page) => void;
+  onFetcherStateChange?: (
+    state: FetcherWithComponents<unknown>["state"]
+  ) => void;
+  projectId: string;
+}) => {
+  const nameFieldId = useId();
+  const pathFieldId = useId();
+
+  const fetcher = useFetcher<CreatePageData>();
+
+  const [fieldErrors, setFieldErrors] = useState<
+    ZodError["formErrors"]["fieldErrors"]
+  >({});
+
+  const isSubmitting = fetcher.state !== "idle";
+
+  const prevFetcher = useRef(fetcher);
+  useEffect(() => {
+    if (prevFetcher.current.state !== fetcher.state) {
+      onFetcherStateChange?.(fetcher.state);
+      if (fetcher.state === "idle" && fetcher.data !== undefined) {
+        if ("ok" in fetcher.data) {
+          onSuccess?.(fetcher.data.page);
+        } else {
```

Comment:
generally prefer early returns and short conditions assigned over variables, makes it easier to read

Reply:
We cannot return early, because we want `prevFetcher.current = fetcher;` at the end to always run.

Reply:
Not sure I understand the idea with `hasFetched`. The way you wrote it will not work for two reasons:

1. TypeScript will not understand that if  `hasFetched` is true, then `fetcher.data` is not undefined
2. If we move `hasFetched` together with `"ok" in fetcher.data`, the else condition will not work as needed

Let's keep as is? Don't see any reason to spend time discussing this really

Reply:
I haven't spent much time in the example, the point I am making is to avoid nested if/else style, because indentation makes it harder to read. Its probably subbjective

Reply:
I get it, but there're reasons to write it this way in this case